lib/auth/keystore/gcp_kms_test.go

```diff
@@ -0,0 +1,367 @@
+	// the next.
+	defer func() {
+		select {
+		case <-m.unblockGetPublicKey:
+		case <-ctx.Done():
+		}
+	}()
+	m.RLock()
+	defer m.RUnlock()
+	if m.keyVersions[req.Name] != "ACTIVE" {
+		return nil, errors.New("GetPublicKey failed, state has value " + m.keyVersions[req.Name])
+	}
+	return &kmspb.PublicKey{
+		Pem: `-----BEGIN PUBLIC KEY-----
+MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAryQICCl6NZ5gDKrnSztO
+3Hy8PEUcuyvg/ikC+VcIo2SFFSf18a3IMYldIugqqqZCs4/4uVW3sbdLs/6PfgdX
+7O9D22ZiFWHPYA2k2N744MNiCD1UE+tJyllUhSblK48bn+v1oZHCM0nYQ2NqUkvS
+j+hwUU3RiWl7x3D2s9wSdNt7XUtW05a/FXehsPSiJfKvHJJnGOX0BgTvkLnkAOTd
+OrUZ/wK69Dzu4IvrN4vs9Nes8vbwPa/ddZEzGR0cQMt0JBkhk9kU/qwqUseP1QRJ
+5I1jR4g8aYPL/ke9K35PxZWuDp3U0UPAZ3PjFAh+5T+fc7gzCs9dPzSHloruU+gl
+FQIDAQAB
+-----END PUBLIC KEY-----`,
+	}, nil
+}
+
+// Never actually succeeds, just blocks until the context is cancelled
+func (m *mockGCPKMSClient) AsymmetricSign(ctx context.Context, req *kmspb.AsymmetricSignRequest, opts ...gax.CallOption) (*kmspb.AsymmetricSignResponse, error) {
+	for {
+		select {
+		case m.signIsBlocked <- struct{}{}:
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+	}
+}
+
+type mockCryptoKeyIterator struct {
+	keys            []string
+	i               int
+	ctx             context.Context
+	unblockIterator chan struct{}
+	sync.Mutex
+}
+
+func (m *mockCryptoKeyIterator) Next() (*kmspb.CryptoKey, error) {
+	select {
+	case <-m.ctx.Done():
+		return nil, m.ctx.Err()
+	case <-m.unblockIterator:
+	}
+	m.Lock()
+	defer m.Unlock()
+	if m.i >= len(m.keys) {
+		return nil, iterator.Done
+	}
+	m.i = m.i + 1
+	return &kmspb.CryptoKey{
+		Name: m.keys[m.i-1],
+	}, nil
+}
+
+// TestGCPKMSKeyStore tests some specific handling of the KMS API. Generic
+// keystore tests are in TestKeyStore.
+func TestGCPKMSKeyStore(t *testing.T) {
+	mockClient := newMockGCPKMSClient()
+	s, err := newGCPKMSKeyStore(&GCPKMSConfig{
+		KeyRing:           "test-keyring",
+		ProtectionLevel:   "HSM",
+		HostUUID:          "test-uuid",
+		kmsClientOverride: mockClient,
+		clockOverride:     mockClient.clock,
+	})
+	require.NoError(t, err)
+
+	// Make sure the keystore can create the key and fetch the public key, even
+	// if the API returns a pending error temporarily.
+	t.Run("generateRSA success", func(t *testing.T) {
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			// Keystore gets a pending key error once...
+			mockClient.unblockGetPublicKey <- struct{}{}
+			mockClient.clock.Advance(defaultGCPPendingRetryInterval)
+
+			// ... and twice.
+			mockClient.unblockGetPublicKey <- struct{}{}
+
+			// Activate the key before advancing the clock so that it can
+			// succeed the third time.
+			mockClient.activateKey <- struct{}{}
+			<-mockClient.keyActivated
+			mockClient.clock.Advance(defaultGCPPendingRetryInterval)
+
+			// It should get the active key here.
+			mockClient.unblockGetPublicKey <- struct{}{}
+		}()
+
+		_, _, err = s.generateRSA()
+		require.NoError(t, err)
```

Comment:
Suggestion: Add an error message to all "require" and "assert" statements that represent what is being tested (instead of test setup) - this hopefully makes eventual errors a bit clearer to future maintainers.